### ISSUE TITLE
Lost redeclaration in the NF

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#9178]
@perost, try to simulate the save total we just sent you. We get something like
```
[1] 15:42:27 Translation Notification[xxx: 4:15-4:86]: From here:
[2] 15:42:27 Translation Error[xxxx: 15:3-15:152]: component source contains the definition of a partial class Medium.
Please redeclare it to any package compatible with Modelica.Media.Interfaces.PartialMedium.
```
From our understanding, the Medium package in the source component and the Medium package in its outlet connector should both be redeclared to non-partial. For some reason, one of the two redeclares is lost.

**Comment by perost** [comment]
I can't find any redeclare of the `Medium` package in the outlet connector, so unless I'm missing something it looks like the error is correct. If I add e.g. `redeclare Medium = Medium` to `C_out` in `FluidSource` and `C_in` in `FluidSink` then the model works as expected.

**Comment by casella** [comment]
As I understand it, the Source model redeclares the outlet connector C_out to WaterSteam.Connectors.Outlet, which is defined as
```
connector Outlet
  extends MetroscopeModelingLibrary.Icons.Connectors.WaterOutletIcon;
  package WaterSteamMedium = MetroscopeModelingLibrary.Media.WaterSteamMedium;
  extends Partial.Connectors.FluidOutlet(redeclare package Medium = WaterSteamMedium) annotation(IconMap(primitivesVisible = false));
end Outlet;
```
the redeclare statement for Medium is in the extends clause. Shouldn't that be enough?

**Comment by perost** [comment]
It seems I was looking at the wrong redeclare, the error message threw me off a bit. The issue wasn't that the redeclare wasn't applied, it was that we checked for partial classes inside of the original components that were being redeclared. I'm not actually sure what makes this model so unique, it looks like a typical fluid model, but the issue should be fixed in #9196.

**Comment by casella** [comment]
@perost, the redeclaration is no longer lost. If we check the `Tests.WaterSteam.Pipes.Pipe_direct` model, the check is successful. However, when trying to simulate it with OMEdit, we get
```
[1] 15:41:28 Translation Error
[XXX.Partial.BoundaryConditions.FluidSource: 22:78-22:91]: Modified element Q not found in class Connectors.
```
This is weird, because element `Q` is of course not contained in class `Connectors`, but rather in class `Connectors.FluidOutlet`, which extends `Connectors.FluidPort`, which includes element `Q`.

**Comment by casella** [comment]
Adding @AndreaBartolini to the loop

**Comment by perost** [comment]
With the new total model I can replicate the issue. The difference is that the `Connectors` package is declared as `partial` in the new model, which somehow causes some issues somewhere. The cause seems to be similar to the one I fixed, that we check for partial elements while instantiating the original component. But I will need to create a minimal example to be able to pinpoint the exact issue.

**Comment by perost** [comment]
> > With the new total model I can replicate the issue. The difference is that the `Connectors` package is declared as `partial` in the new model
> 
> Yes, but then it is redeclared.

Yes, I'm just pointing out what the difference between the models is.

> > that we check for partial elements while instantiating the original component. But I will need to create a minimal example to be able to pinpoint the exact issue.
> 
> OK, let us know if you need help with that. Thanks!

It's fine, I've made so many MWE:s that I could probably do it in my sleep :smile:

**Comment by perost** [comment]
Fixed in #9209.

The new model now works, but in my opinion it probably shouldn't. The model contains e.g. `Source source` which redeclares `C_out` as `Outlet`, and `Outlet` extends from `Partial.Connectors.FluidOutlet` where `Connectors` is partial (and `Connectors` is never redeclared anywhere in the total model).

I don't think this should be allowed, but the check for this currently looks like:
https://github.com/OpenModelica/OpenModelica/blob/7e822a28092bb9a5c4a6de6b1cf2b63049340138/OMCompiler/Compiler/NFFrontEnd/NFLookup.mo#L691-L696

I think this is only an issue in some older Modelica versions like 3.1 or something, so we might at some point look into enabling it for newer versions. So it would probably be best to avoid using a partial package in this way or it might cause issues in the future.

**Comment by perost** [comment]
> > Like I said I think it's only needed for some old versions of the MSL, so we might be able to enable it for newer versions of Modelica. That will probably have to wait until after my vacation, but I'll open an issue for it so we don't forget.
> 
> As far as I am concerned, MSL 3.1 is not really S, i.e. it was demonstrably not compliant with the standard. To be honest, I don't care much what OMC says when it is used. BTW, MSL 3.2.3 is backwards-compatible with it, so there is really no reason to keep using MSL 3.1 in OMC. I would not worry too much about that.

My mistake, it was actually only fixed in MSL 3.2.3 (see [#3615](https://github.com/modelica/ModelicaStandardLibrary/issues/3615)). That makes it a bit complicated because it means we can't just use the Modelica version to conditionally enable the check, since both e.g. 3.2.3 and 3.2.2 uses Modelica 3.2.

Your point about 3.2.3 being backward-compatible still stands of course, but I think we should still take a look at whether we can avoid breaking things unnecessarily. Maybe just add a workaround specifically for `PartialModelicaServices` or patch our own libraries, the fix is just to make `PartialModelicaServices` not partial after all. Just enabling the check will break some of our test cases and probably most of the library testing of pre-MSL 3.2.3, so it will require a bit of work regardless.

**Comment by AndreaBartolini** [comment]
Hi per,
thanks for the analysis, I'll investigate deeply in the library in order to fix this kind of issues.
I agree with you on the fact that there is no reason to made "partial" the packages inside that library... we are involved in the library audit mainly to find and solve issues like this.

Have a nice holidays,
Andrea
